File: Python/Leetcode/Medium/Group the People Given the Group Size They Belong To.py
from collections import defaultdict
from copy import deepcopy

class Solution:
    def groupThePeople(self, groupSizes: list[int]) -> list[list[int]]:

        result = [] # 결과에 담을 변수
        length = len(groupSizes)
        map = defaultdict(list)

        for i in range(length):
            size = groupSizes[i]
            # map에 키값이 없을 경우 새로 추가 한다.
            if size not in map.keys():
                map[size] = []

            map[size].append(i)

            # map value와 size와 같을 경우 결과 리스트에 추가하고 value 리스트를 초기화 한다.
            if len(map[size]) == size:
                result.append(deepcopy(map[size]))
                map[size].clear()

        return result

solution = Solution()
print(solution.groupThePeople([3,3,3,3,3,1,3]))
print(solution.groupThePeople([2,1,3,3,3,2]))

File: Java/Leetcode/src/Medium/GroupThePeopleGivenTheGroupSizeTheyBelongTo.java
package Medium;

import java.util.*;

public class GroupThePeopleGivenTheGroupSizeTheyBelongTo {

    public List<List<Integer>> groupThePeople(int[] groupSizes) {

        List<List<Integer>> result = new ArrayList<>(); // 결과
        int length  = groupSizes.length;
        Map<Integer, List<Integer>> map = new HashMap<>(); // map

        int size;
        for(int i = 0; i < length; i++){
            size = groupSizes[i];
            // Map에 없을 경우 리스트 추가
            if(!map.containsKey(size)){
                map.put(size, new ArrayList<>());
            }

            map.get(size).add(i);

            // Group Size와 리스트 길이 같을때 Result에 저장
            if(map.get(size).size() == size){
                result.add(new ArrayList<>(map.get(size)));
                map.get(size).clear();
            }
        }

        return result;
    }

    public static void main(String[] args) {
        GroupThePeopleGivenTheGroupSizeTheyBelongTo obj = new GroupThePeopleGivenTheGroupSizeTheyBelongTo();
        int[] groupSizes = {3,3,3,3,3,1,3};
        System.out.println(obj.groupThePeople(groupSizes));
        groupSizes = new int[] {2,1,3,3,3,2};
        System.out.println(obj.groupThePeople(groupSizes));

    }

}
